### PR TITLE
Fix "failed to create process" issue on Windows

### PR DESCRIPTION
**Pull request by fkrull**
Fix for issue #398. As I explained there (https://github.com/pypa/setuptools/issues/398#issuecomment-225444249), this is actually a setuptools issue. I added a test to check for the problem as well as the fix itself.

This seems to be only relevant to Windows; according to https://lists.gnu.org/archive/html/bug-bash/2008-05/msg00052.html, "Unix" (whatever that means in the context) doesn't have a way to escape spaces in the interpreter name so I didn't test for that.

### CODE REVIEW

**Comment by jaraco** [comment]
I very much appreciate the test and the implementation. How confident can we be that this fix doesn't cause a regression w.r.t. #188?

**Comment by fkrull** [comment]
...that's a good point, actually; it does regress on that.

It's tricky because the 'executable' argument that install_scripts pulls from build_scripts can be either
1. a single command with spaces in its path (reasonably likely on Windows, not supported on Unix)
2. or a command plus arguments (possible on both).
And there's no way to tell which it is. The current behaviour always assumes 2, this PR right now always assumes 1 which breaks issue #188 again.

Something has to give. My preference would be 1 on Windows, 2 everywhere else; that makes it impossible to specify extra arguments on Windows, but Python installed in Program Files is more useful IMHO.

**Comment by fkrull** [comment]
Ok, I went back and forth on it a bit and I think I came up with a better solution by only special-casing sys.executable. I added some more tests that check shebang quoting for Windows and Unix. I also ended up doing some simple test maintenance, so now all tests succeed on Windows with Python installed in Program Files (they didn't before).

**Comment by fkrull** [comment]
Bump?

**Comment by jaraco** [comment]
Sorry it's taken so long to get to this. And thanks for the ping.

Overall, this seems like a sound solution (much better than just vacillating between two broken implementations).

I'll make some comments inline to the commits.

**Comment by jaraco** [comment]
Also, I appreciate the extra effort that you've put into this, given that the status quo is that the last change broke the tests on Windows. It's a shame that the tests haven't been run on Windows frequently enough to catch the former regressions.

File: setuptools/command/install_scripts.py [hunk 1]
```diff
@@ -37,6 +38,10 @@ def run(self):
         if is_wininst:
             exec_param = "python.exe"
             writer = ei.WindowsScriptWriter
+        if exec_param == sys.executable:
```
Comment (jaraco) [inline setuptools/command/install_scripts.py]: The whole point of creating the CommandSpec class was to try to encapsulate as much of these decision points as possible, so it's sad to see this one re-introduced. What's worse is that it seems to be special-casing a special case, inferring that if the string is `sys.executable`, then it was meant to be escaped, but in any other case it's not. This expectation would do the wrong thing if someone passed an executable with spaces that wasn't sys.executable (e.g. "C:\My Python Env\python.exe" where sys.executable is "C:\Python35\python.exe"). I'm not sure if there's something better do be done other than grin and bear it.

File: setuptools/tests/test_install_scripts.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        cmd = install_scripts(dist)
+        cmd.install_dir = install_dir
+        if executable is not None:
+            bs = cmd.get_finalized_command('build_scripts')
+            bs.executable = executable
+        cmd.ensure_finalized()
+        with contexts.quiet():
+            cmd.run()
+
+    @pytest.mark.skipif(sys.platform == 'win32', reason='non-Windows only')
```
Comment (jaraco) [inline setuptools/tests/test_install_scripts.py]: I dislike seeing all of these platform-specific tests. The apparent dichotomy here suggests that there are actually two behaviors of setuptools, one for Windows and another for other. Perhaps that makes sense, but as much as possible, I'd like for the behavior to be universal and tested on both platforms. That way, even though the behavior wouldn't be invoked in that form on that platform in the wild, it'll be tested by someone developing setuptools on that platform. Still, something that we can iterate on subsequently.

**Comment by fkrull** [comment]
Regarding the `sys.executable` special-case: I couldn't come up with a way that didn't break in *some* situation. The optimal solution would probably be to extend build_scripts and handle the --executable argument in such a way that the user can specify a string with spaces or a list of strings (possibly with spaces). I'm not familiar enough with distutils to know if its argument parsing allows that or if that can be done in a backwards-compatible way.

**Pull request merged**